Storage/SQL/SQLiteHistoryFavicons.swift

```diff
@@ -0,0 +1,439 @@
+        let deferred = CancellableDeferred<Maybe<HTMLDocument>>()
+
+        alamofire.request(url).response { response in
+            guard response.error == nil,
+                let data = response.data,
+                let document = try? HTMLDocument(data: data) else {
+                deferred.fill(Maybe(failure: FaviconLookupError(siteURL: url.absoluteString)))
+                return
+            }
+
+            deferred.fill(Maybe(success: document))
+        }
+
+        return deferred
+    }
+
+    // Scrapes the web page at the specified URL for its favicon URLs.
+    fileprivate func getFaviconURLsFromWebPage(url: URL) -> Deferred<Maybe<[URL]>> {
+        return getHTMLDocumentFromWebPage(url: url).bind { result in
+            guard let document = result.successValue else {
+                return deferMaybe(FaviconLookupError(siteURL: url.absoluteString))
+            }
+
+            // If we were redirected via a <meta> tag on the page to a different
+            // URL, go to the redirected page for the favicon instead.
+            for meta in document.xpath("//head/meta") {
+                if let refresh = meta["http-equiv"], refresh == "Refresh",
+                    let content = meta["content"],
+                    let index = content.range(of: "URL="),
+                    let reloadURL = URL(string: String(content[index.upperBound...])),
+                    reloadURL != url {
+                    return self.getFaviconURLsFromWebPage(url: reloadURL)
+                }
+            }
+
+            var icons = [URL]()
+
+            // Iterate over each <link rel="icon"> tag on the page.
+            for link in document.xpath("//head//link[contains(@rel, 'icon')]") {
+                // Only consider <link rel="icon"> tags with an [href] attribute.
+                if let href = link["href"], let faviconURL = URL(string: href, relativeTo: url) {
+                    icons.append(faviconURL)
+                }
+            }
+
+            // Also, consider a "/favicon.ico" icon at the root of the domain.
+            if let faviconURL = URL(string: "/favicon.ico", relativeTo: url) {
+                icons.append(faviconURL)
+            }
+
+            return deferMaybe(icons)
+        }
+    }
+
+    // Generates a "default" favicon based on the first character in the
+    // site's domain name or gets an already-generated icon from the cache.
+    fileprivate func getDefaultFaviconImage(forSite site: Site) -> Deferred<Maybe<UIImage>> {
+        guard let url = URL(string: site.url),
+            let character = url.baseDomain?.first else {
+            return deferMaybe(defaultFavicon)
+        }
+
+        let faviconLetter = String(character).uppercased()
+
+        if let cachedFavicon = defaultFaviconImageCache[faviconLetter] {
+            return deferMaybe(cachedFavicon)
+        }
+
+        func getDefaultColor(_ url: URL) -> UIColor {
```

Comment:
nit: generateBackgroundColor(fromURL url: URL)

Reply:
It is `getter-like` internally, but I think specifying this as if the color is 'generated' seems clearer to me when reading the code.